woocommerce/woocommerce-gutenberg-products-block
# Add minimum quantity, maximum quantity, and step (multiple_of) to the Cart Block and Store API

Closes https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/5037

Implements the following changes:

1. `quantity_limits` returned from the Store API `cart` endpints, which contains `minimum`, `maximum`, and `multiple_of`. There is also an `editable` prop which can be set to false to disable inputs.
2. For products, which are not yet in the cart remember, the product route returns `minimum`, `maximum`, and `multiple_of` as part of the `add_to_cart` param. I've updated the `QuantityInput` component to match and support step, in the same way @senadir added support to `QuantitySelector`.
3. `quantity_limits` are validated by the cart controller when adding an item to the cart, or when updating an item already in the cart. The API returns error 400 if the provided value is out of bounds.
4. ~~Styles the quantity input for stepped (`multiple_of`) quantities so the +/- buttons clearly show what value comes next.~~
5. Hides the quantity selector if the product is sold individually or if `editable` is false.

Fixes #3888

#### Caveats:

This functionality is implemented in the Store API. Therefore, minimum, maximum, and step attributes can be circumvented via core functionality e.g. Shop page. Core offers its own filters to control this.

#### List of hooks

- `woocommerce_store_api_product_quantity_multiple_of`
- `woocommerce_store_api_product_quantity_minimum`
- `woocommerce_store_api_product_quantity_maximum`
- `woocommerce_store_api_product_quantity_editable` (cart only)

### Testing steps

#### Developer testing steps:

Smoke test adding items to cart and changing quantity, either via the + and - buttons, or directly via the input.

**Disable Selectors**

Add the following code to disable quantity inputs for cart items:

```php
add_filter( 'woocommerce_store_api_product_quantity_editable', function( $quantity, $product, $cart_item )  {
	return false;
}, 10, 3 );
```

Test that no selectors are shown on the cart page. If you attempt to update a qty via the API for an existing cart item, you will get a 400 error.

**Step Property**

Add the following code to change the multiple/step property for all cart items:

```php
add_filter( 'woocommerce_store_api_product_quantity_multiple_of', function( $quantity, $product, $cart_item )  {
	return 4;
}, 10, 3 );
```

- Refresh your cart, change values, quantity should up and down by 4 (or whatever value you set above).
- If the initial value isn't a multiplier of 4, after changing the quantity, your new quantity would be a multiplier of 4.
- Try to type a number inside the quantity selector, trying something that isn't a multiplier of 4 would change to the nearest one, for example, typing 14 would yield back 12.

**Min/Max for Cart Items**

Add the following code to change the minimum and maximum for all cart items:

```php
add_filter( 'woocommerce_store_api_product_quantity_minimum', function( $quantity, $product, $cart_item )  {
	return 5;
}, 10, 3 );

add_filter( 'woocommerce_store_api_product_quantity_maximum', function( $quantity, $product, $cart_item )  {
	return 30;
}, 10, 3 );
```

- Refresh your cart, change values. You'll be restricted to a qty between 5 and 30.

**Min/Max for Cart Products**

Add the following code to change the minimum qty for products (this affects add to cart events):

```php
add_filter( 'woocommerce_store_api_product_quantity_minimum', function( $quantity, $product, $cart_item )  {
	return 5;
}, 10, 3 );
```

- Use the All Products block to add a product to the cart.
- You should see 5 in cart after adding to the cart the first time.

#### User facing testing steps:

1. Smoke test adding items to cart and changing quantity, either via the + and - buttons, or directly via the input.
2. Edit a product and make it "sold individually"
3. Confirm that no quantity box shows on the cart page for this product
4. Edit a product and set stock to 6, no backorders.
5. Confirm that you can only have 6 maximum for this product.

#### Screenshots

Qty vs qty disabled:

![Screenshot 2021-12-21 at 14 39 20](https://user-images.githubusercontent.com/90977/146947952-208e8fee-630b-4258-b101-392f6b575e04.png)
![Screenshot 2021-12-21 at 14 39 16](https://user-images.githubusercontent.com/90977/146947955-1c145d23-92bc-4657-bddc-21fd833e4c31.png)

#### Changelog

> Store API and Cart block now support defining a quantity stepper and a minimum quantity.
